File: volterra-braid/oracle/analyse_run.py
#!/usr/bin/env python3
"""Extract the braid of the mobile +1/2 defects from an `fd` run directory.

Generalises `extract_braid.py` from the steady-winding circle to any
confinement the `fd` driver can build. Two things change with the geometry:

- The interior mask is read from the run's own `mask.txt` rather than
  recomputed. An epitrochoid's interior test is a root solve per cell, so
  recomputing it here would mean porting that solve into Python.
- Defects are split by charge, and only the positive ones enter the braid. On
  the circle every defect in the braiding regime is a +1/2, so the split is a
  no-op there. On an epitrochoid each regularised cusp pins a -1/2 defect,
  which is stationary, sits against the boundary, and is not a braid strand.

Usage:
    python analyse_run.py <run_dir> [threshold] [--min-window N] [--winding]

`run_dir` is the directory holding `Q/`, `mask.txt` and `meta.json`.

`--winding` swaps the published Jacobian-and-threshold detector for plaquette
winding of the director, which carries no threshold. A threshold calibrated at
one coherence length is wrong at another, and every epitrochoid point in
arXiv:2503.10880 Fig. 7 sits an order of magnitude in coherence length away from
the circle runs the published threshold of 0.1 was set for. This implementation
is deliberately independent of `volterra_braid::detect_defects_winding`, so
running both on the same frames checks one against the other.
"""
import glob
import json
import sys
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parent))
import braid_tracker_v2 as v2  # noqa: E402


def load_frame(path):
    data = np.loadtxt(path)
    return data[:, 0], data[:, 1]


def detect_defects_winding(qxx, qxy, nx, ny, mask):
    """Defects by the winding of the director around each lattice plaquette.

    Returns (x, y, charge) triples with charge +/-1, in the same convention as
    ``braid_tracker_v2.detect_defects``. Vectorised over plaquettes, so it is
    not a transcription of the Rust and disagreement means one of them is wrong.
    """
    qxx = np.asarray(qxx, float).reshape(nx, ny)
    qxy = np.asarray(qxy, float).reshape(nx, ny)
    mask = np.asarray(mask, bool).reshape(nx, ny)

    phi = 0.5 * np.arctan2(qxy, qxx)

    def wrap_half(d):
        # Into (-pi/2, pi/2]: the director is a line, so a corner-to-corner turn
        # past a right angle reads as the shorter turn the other way.
        return -((-d + np.pi / 2) % np.pi - np.pi / 2)

    corners = [phi[:-1, :-1], phi[1:, :-1], phi[1:, 1:], phi[:-1, 1:]]
    total = sum(wrap_half(corners[(i + 1) % 4] - corners[i]) for i in range(4))
    inside = mask[:-1, :-1] & mask[1:, :-1] & mask[1:, 1:] & mask[:-1, 1:]
    charge = np.where(inside, np.rint(total / np.pi), 0.0).astype(int)

    # 8-connected same-sign clusters, centroid of plaquette centres.
    visited = np.zeros_like(charge, bool)
    neigh = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]
    out = []
    px, py = charge.shape
    for x in range(px):
        for y in range(py):
            if charge[x, y] == 0 or visited[x, y]:
                continue
            sign = int(np.sign(charge[x, y]))
            stack = [(x, y)]
            visited[x, y] = True
            sx = sy = 0.0
            count = 0
            while stack:
                cx, cy = stack.pop()
                sx += cx + 0.5
                sy += cy + 0.5
                count += 1
                for dx, dy in neigh:
                    a, b = cx + dx, cy + dy
                    if not (0 <= a < px and 0 <= b < py):
                        continue
                    if not visited[a, b] and charge[a, b] != 0 and np.sign(charge[a, b]) == sign:
                        visited[a, b] = True
                        stack.append((a, b))
            out.append((sx / count, sy / count, sign))
    return out


def trailing_runs(counts):
    """Maximal runs of a constant count, latest first."""
    runs = []
    i = len(counts) - 1
    while i >= 0:
        j = i
        while j - 1 >= 0 and counts[j - 1] == counts[i]:
            j -= 1
        runs.append((j, i, counts[i]))
        i = j - 1
    return runs


def fmt(codes):
    return " ".join(
        (f"sigma_{abs(g)}^-1" if g < 0 else f"sigma_{g}") for g in codes
    )


def main():
    run_dir = Path(sys.argv[1])
    threshold = 0.1
    min_window = 8
    rest = sys.argv[2:]
    if rest and not rest[0].startswith("--"):
        threshold = float(rest[0])
        rest = rest[1:]
    if "--min-window" in rest:
        min_window = int(rest[rest.index("--min-window") + 1])
    winding = "--winding" in rest

    meta = json.loads((run_dir / "meta.json").read_text())
    lx, ly = meta["lx"], meta["ly"]

    mask_path = run_dir / "mask.txt"
    if not mask_path.exists():
        print(f"no mask.txt under {run_dir}; the run predates mask output")
        sys.exit(2)
    mask = np.loadtxt(mask_path).astype(bool)
    if mask.size != lx * ly:
        print(f"mask has {mask.size} cells, grid is {lx}x{ly}")
        sys.exit(2)

    files = sorted(glob.glob(str(run_dir / "Q" / "Q_*.txt")))
    if not files:
        print(f"no Q frames under {run_dir}/Q")
        sys.exit(2)

    positive, negative = [], []
    for f in files:
        qxx, qxy = load_frame(f)
        if winding:
            found = detect_defects_winding(qxx, qxy, lx, ly, mask)
        else:
            found = v2.detect_defects(qxx, qxy, lx, ly, threshold, mask)
        positive.append([d for d in found if d[2] > 0])
        negative.append([d for d in found if d[2] < 0])

    pos_counts = [len(d) for d in positive]
    neg_counts = [len(d) for d in negative]
    print(f"{len(files)} frames over {meta['n_steps']} steps, "
          f"{meta['lx']}x{meta['ly']}, als={meta['als']:.4f} ncl={meta['ncl']:.4f}")
    print(f"  +1/2 counts: {pos_counts}")
    print(f"  -1/2 counts: {neg_counts}")

    # Where the pinned negatives sit, averaged over the trailing half of the
    # run. A cusp-pinned defect barely moves, so a large spread means it is not
    # pinned.
    tail = negative[len(negative) // 2:]
    if any(tail):
        pts = np.array([[d[0], d[1]] for frame in tail for d in frame])
        centre = (lx / 2 - 1, ly / 2 - 1)
        radii = np.hypot(pts[:, 0] - centre[0], pts[:, 1] - centre[1])
        print(f"  -1/2 radial position over the trailing half: "
              f"mean {radii.mean():.1f}, sd {radii.std():.1f} px "
              f"(domain radius ~{lx / 2 - 1:.0f})")

    runs = trailing_runs(pos_counts)
    chosen = next(
        ((s, e, c) for (s, e, c) in runs if e - s + 1 >= min_window and c >= 2),
        None,
    )
    if chosen is None:
        print(f"no run of >={min_window} frames with a stable +1/2 count and "
              f"at least 2 defects")
        sys.exit(3)

    start, end, count = chosen
    print(f"  braiding frames [{start}, {end}] ({end - start + 1} frames), "
          f"{count} mobile +1/2 defects")

    n, codes = v2.braidword_from_frames(positive[start:end + 1])
    period = v2.period_word(codes) if codes else []
    h_period = v2.topological_entropy(n, period) if period else 0.0
    h_window = v2.topological_entropy(n, codes) if codes else 0.0
    repeats = len(codes) / len(period) if period else 0.0

    print(f"n_strands={n}")
    print(f"window word ({len(codes)} generators, {repeats:.2f} periods)")
    print(f"braid word: {{{fmt(period)}}}" if period else "braid word: {} (trivial)")
    print(f"topological entropy: {h_period:.6f}")
    print(f"whole-window entropy (not comparable to a published value): {h_window:.6f}")


if __name__ == "__main__":
    main()

File: volterra-braid/examples/analyse_run.rs
//! Read an `fd` run directory and report the braid of its mobile `+1/2` defects.
//!
//! ```text
//! cargo run --release -p volterra-braid --example analyse_run -- <run_dir> [--min-window N]
//! ```
//!
//! `<run_dir>` is what the `fd` driver writes: `Q/Q_*.txt` frames and a
//! `mask.txt` giving the confined interior. Detection is by director winding
//! ([`volterra_braid::detect_defects_winding`]), which carries no threshold, so
//! the same command works for a steady-winding circle whose defect cores are
//! under a lattice spacing across and for an epitrochoid whose cores are ten.
//!
//! Negative defects are reported but excluded from the braid. On an epitrochoid
//! each regularised cusp pins a `-1/2` defect, which is part of the effective
//! boundary rather than a braid strand.

use std::collections::BTreeMap;
use std::error::Error;
use std::fs;
use std::path::{Path, PathBuf};

use volterra_braid::{
    BraidWord, Defect, Worldline, detect_defects_winding, extract_braidword, track,
};

fn read_mask(path: &Path) -> Result<Vec<bool>, Box<dyn Error>> {
    Ok(fs::read_to_string(path)?
        .split_whitespace()
        .map(|t| t != "0")
        .collect())
}

fn read_frame(path: &Path) -> Result<(Vec<f64>, Vec<f64>), Box<dyn Error>> {
    let text = fs::read_to_string(path)?;
    let mut qxx = Vec::new();
    let mut qxy = Vec::new();
    for line in text.lines() {
        let mut it = line.split_whitespace();
        let (Some(a), Some(b)) = (it.next(), it.next()) else {
            continue;
        };
        // Naming the file matters: pointed at a run still in flight, this hits
        // the frame currently being written, and the failure should say so
        // rather than read as a diverged field.
        let parse = |t: &str| -> Result<f64, Box<dyn Error>> {
            t.parse::<f64>()
                .map_err(|e| format!("{}: {e} on {t:?}", path.display()).into())
        };
        qxx.push(parse(a)?);
        qxy.push(parse(b)?);
    }
    Ok((qxx, qxy))
}

/// Maximal runs of a constant count, latest first, as `(start, end, count)`.
fn trailing_runs(counts: &[usize]) -> Vec<(usize, usize, usize)> {
    let mut runs = Vec::new();
    let mut i = counts.len() as i64 - 1;
    while i >= 0 {
        let mut j = i;
        while j > 0 && counts[(j - 1) as usize] == counts[i as usize] {
            j -= 1;
        }
        runs.push((j as usize, i as usize, counts[i as usize]));
        i = j - 1;
    }
    runs
}

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = std::env::args().collect();
    let Some(run_dir) = args.get(1).map(PathBuf::from) else {
        eprintln!("usage: analyse_run <run_dir> [--min-window N]");
        std::process::exit(2);
    };
    let min_window: usize = args
        .iter()
        .position(|a| a == "--min-window")
        .and_then(|i| args.get(i + 1))
        .and_then(|v| v.parse().ok())
        .unwrap_or(8);

    let mask = read_mask(&run_dir.join("mask.txt"))?;
    let n = mask.len();
    let lx = (n as f64).sqrt().round() as usize;
    if lx * lx != n {
        return Err(format!("mask has {n} cells, which is not a square grid").into());
    }
    let ly = lx;

    let mut frames: Vec<PathBuf> = fs::read_dir(run_dir.join("Q"))?
        .filter_map(|e| e.ok().map(|e| e.path()))
        .filter(|p| p.extension().is_some_and(|e| e == "txt"))
        .collect();
    frames.sort();
    if frames.is_empty() {
        return Err(format!("no Q frames under {}", run_dir.join("Q").display()).into());
    }

    let mut positive: Vec<Vec<Defect>> = Vec::with_capacity(frames.len());
    let mut negative: Vec<Vec<Defect>> = Vec::with_capacity(frames.len());
    for (i, f) in frames.iter().enumerate() {
        let (qxx, qxy) = match read_frame(f) {
            Ok(v) => v,
            // Only the newest frame is allowed to be unreadable, and only
            // because pointing this at a run still in flight catches the frame
            // being written. Anywhere earlier, an unreadable frame is a
            // corrupted one and the run should be looked at, not analysed.
            Err(e) if i + 1 == frames.len() => {
                println!("  dropping the newest frame, still being written: {e}");
                frames.pop();
                break;
            }
            Err(e) => return Err(e),
        };
        let found = detect_defects_winding(&qxx, &qxy, lx, ly, &mask);
        positive.push(found.iter().copied().filter(|d| d.charge > 0).collect());
        negative.push(found.into_iter().filter(|d| d.charge < 0).collect());
    }

    let pos_counts: Vec<usize> = positive.iter().map(Vec::len).collect();
    let neg_counts: Vec<usize> = negative.iter().map(Vec::len).collect();
    println!(
        "{} frames on a {lx}x{ly} grid, {} interior cells, sqrt(A_sys)={:.3}",
        frames.len(),
        mask.iter().filter(|&&b| b).count(),
        (mask.iter().filter(|&&b| b).count() as f64).sqrt()
    );

    // Counts as a histogram rather than a per-frame list: at several hundred
    // frames the list is unreadable and what matters is which count dominates.
    let histogram = |counts: &[usize]| {
        let mut h: BTreeMap<usize, usize> = BTreeMap::new();
        for &c in counts {
            *h.entry(c).or_default() += 1;
        }
        h.iter()
            .map(|(c, n)| format!("{c}x{n}"))
            .collect::<Vec<_>>()
            .join(" ")
    };
    println!("  +1/2 per frame (count x frames): {}", histogram(&pos_counts));
    println!("  -1/2 per frame (count x frames): {}", histogram(&neg_counts));

    // Where the negatives sit over the trailing half. A cusp-pinned defect
    // barely moves, so a large spread is evidence against pinning.
    let tail = &negative[negative.len() / 2..];
    let centre = (lx as f64 / 2.0 - 1.0, ly as f64 / 2.0 - 1.0);
    let radii: Vec<f64> = tail
        .iter()
        .flatten()
        .map(|d| ((d.pos[0] - centre.0).powi(2) + (d.pos[1] - centre.1).powi(2)).sqrt())
        .collect();
    if !radii.is_empty() {
        let mean = radii.iter().sum::<f64>() / radii.len() as f64;
        let sd = (radii.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / radii.len() as f64).sqrt();
        println!(
            "  -1/2 radial position over the trailing half: mean {mean:.1}, sd {sd:.1} px \
             (domain radius ~{:.0})",
            lx as f64 / 2.0 - 1.0
        );
    }

    let transient = frames.len() / 4;

    // Two ways to choose the frames the braid is read from.
    //
    // The default is the longest stable-count stretch past the initial
    // transient, rather than the latest one: the last stable stretch can be a
    // short tail that happens to end the run. Ties go to the later stretch.
    //
    // --bridge-interruptions instead keeps every post-transient frame whose
    // count equals the modal count and drops the rest. arXiv:2503.10880 reports
    // most epitrochoid braids as interrupted, each +1/2 defect being
    // pair-annihilated at a cusp and re-emitted at the same place a moment
    // later, and rectifies this by "permitting that the newly created defect
    // carries on the same braid strands as the recently annihilated defect".
    // Dropping the frames where the count dips and letting the tracker resume
    // is that rectification: the nearest previous strand to a defect emitted at
    // a cusp is the one that vanished into it. The count of dropped frames is
    // printed, since the rectification is only defensible while it stays small.
    let (window, start, end, count): (Vec<Vec<Defect>>, usize, usize, usize) =
        if args.iter().any(|a| a == "--bridge-interruptions") {
            let mut modal: BTreeMap<usize, usize> = BTreeMap::new();
            for &c in &pos_counts[transient..] {
                if c >= 2 {
                    *modal.entry(c).or_default() += 1;
                }
            }
            let Some((&c, &held)) = modal.iter().max_by_key(|&(_, n)| *n) else {
                println!("no post-transient frame carries two or more +1/2 defects");
                std::process::exit(3);
            };
            let kept: Vec<usize> = (transient..pos_counts.len())
                .filter(|&i| pos_counts[i] == c)
                .collect();
            let dropped = pos_counts.len() - transient - kept.len();
            println!(
                "  bridging interruptions: {held} frames at the modal count {c}, \
                 {dropped} dropped ({:.0}% of the post-transient run)",
                100.0 * dropped as f64 / (pos_counts.len() - transient) as f64
            );
            (
                kept.iter().map(|&i| positive[i].clone()).collect(),
                kept[0],
                *kept.last().unwrap(),
                c,
            )
        } else {
            let mut candidates: Vec<(usize, usize, usize)> = trailing_runs(&pos_counts)
                .into_iter()
                .filter(|&(s, e, c)| e >= transient && e - s + 1 >= min_window && c >= 2)
                .collect();
            candidates.sort_by_key(|&(s, e, _)| (e - s, e));
            let Some(&(s, e, c)) = candidates.last() else {
                println!("no run of >={min_window} frames with a stable +1/2 count of at least 2");
                std::process::exit(3);
            };
            (positive[s..=e].to_vec(), s, e, c)
        };
    println!(
        "  braiding frames [{start}, {end}] ({} frames), {count} mobile +1/2 defects",
        window.len()
    );

    let window = &window[..];
    report_period(window);

    // The braid word depends on the projection direction; the topological
    // entropy of a genuinely pseudo-Anosov stirring does not, as long as the
    // projection is generic. A degenerate direction, one along which two
    // strands stay nearly collinear, drops crossings and reads low. Scanning
    // the angle separates the two: a real braid holds its entropy over a range
    // of angles, and an artefact of one projection does not.
    if args.iter().any(|a| a == "--axis-scan") {
        println!("  projection scan (angle, period length, entropy per period):");
        for k in 0..12 {
            let theta = std::f64::consts::PI * k as f64 / 12.0;
            let rotated: Vec<Vec<Defect>> = window.iter().map(|f| rotate(f, theta)).collect();
            let w = extract_braidword(&track(&rotated));
            let p = w.period_word();
            println!(
                "    {:>5.1} deg  {:>3} gens  h = {:.6}  {{{}}}",
                theta.to_degrees(),
                p.gens.len(),
                w.entropy_per_period(),
                format_word(&p)
            );
        }
    }

    let worldlines: Vec<Worldline> = track(window);
    let word: BraidWord = extract_braidword(&worldlines);
    let period = word.period_word();

    println!("n_strands={}", word.n_strands);
    println!(
        "window word ({} generators, {:.2} periods)",
        word.gens.len(),
        if period.gens.is_empty() {
            0.0
        } else {
            word.gens.len() as f64 / period.gens.len() as f64
        }
    );
    if period.gens.is_empty() {
        println!("braid word: {{}} (trivial)");
    } else {
        println!("braid word: {{{}}}", format_word(&period));
    }
    println!("topological entropy: {:.6}", word.entropy_per_period());
    println!(
        "whole-window entropy (not comparable to a published value): {:.6}",
        word.topological_entropy()
    );
    Ok(())
}

fn format_word(w: &BraidWord) -> String {
    w.gens
        .iter()
        .map(|g| {
            if g.inverse {
                format!("sigma_{}^-1", g.index)
            } else {
                format!("sigma_{}", g.index)
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

/// Rotate a frame's defect positions by `theta` about the origin.
///
/// `extract_braidword` projects onto the first coordinate, so rotating the
/// positions is how the projection direction is chosen. Tracking is by distance
/// and so is unaffected.
fn rotate(frame: &[Defect], theta: f64) -> Vec<Defect> {
    let (s, c) = theta.sin_cos();
    frame
        .iter()
        .map(|d| Defect {
            pos: [
                c * d.pos[0] + s * d.pos[1],
                -s * d.pos[0] + c * d.pos[1],
            ],
            charge: d.charge,
        })
        .collect()
}

/// Report the dominant period of the defect configuration, in frames.
///
/// Braid extraction is only as good as the sampling: a braid word is read from
/// the order of the strands changing, so a run sampled at four frames per
/// period cannot resolve a period carrying four generators, however long the
/// run is. This measures the period without tracking anything, so it is
/// independent of the extraction it is there to qualify.
///
/// The observable is the mean pairwise distance among the `+1/2` defects, which
/// is invariant under relabelling them and so needs no worldlines. Its
/// autocorrelation peaks at the period.
fn report_period(window: &[Vec<Defect>]) {
    let spread: Vec<f64> = window
        .iter()
        .map(|frame| {
            let mut total = 0.0;
            let mut pairs = 0usize;
            for i in 0..frame.len() {
                for j in i + 1..frame.len() {
                    let dx = frame[i].pos[0] - frame[j].pos[0];
                    let dy = frame[i].pos[1] - frame[j].pos[1];
                    total += (dx * dx + dy * dy).sqrt();
                    pairs += 1;
                }
            }
            if pairs == 0 { 0.0 } else { total / pairs as f64 }
        })
        .collect();
    if spread.len() < 12 {
        return;
    }
    let mean = spread.iter().sum::<f64>() / spread.len() as f64;
    let dev: Vec<f64> = spread.iter().map(|v| v - mean).collect();
    let norm: f64 = dev.iter().map(|v| v * v).sum();
    if norm <= 0.0 {
        println!("  defect configuration is stationary: no period to resolve");
        return;
    }
    let max_lag = (spread.len() / 3).min(200);
    let mut best = (0usize, 0.0f64);
    for lag in 2..max_lag {
        let r: f64 = (0..dev.len() - lag).map(|i| dev[i] * dev[i + lag]).sum::<f64>() / norm;
        if r > best.1 {
            best = (lag, r);
        }
    }
    let rel_spread = (norm / spread.len() as f64).sqrt() / mean;
    if best.1 < 0.2 {
        println!(
            "  no clear period: strongest autocorrelation {:.2} at lag {} frames, \
             configuration varying by {:.0}% of its mean",
            best.1,
            best.0,
            100.0 * rel_spread
        );
    } else {
        println!(
            "  period about {} frames (autocorrelation {:.2}), configuration varying by \
             {:.0}% of its mean",
            best.0,
            best.1,
            100.0 * rel_spread
        );
    }
}
